src/nmodl/solver/matexp/MatrixExponential.h

```diff
@@ -0,0 +1,483 @@
+        } else if (l1norm < 1.880152677804762e+000f) {
+            matrix_exp_pade5(arg, U, V);
+        } else {
+            const float maxnorm = 3.925724783138660f;
+            frexp(l1norm / maxnorm, &squarings);
+            if (squarings < 0)
+                squarings = 0;
+            MatrixType A = arg.unaryExpr(MatrixExponentialScalingOp<float>(squarings));
+            matrix_exp_pade7(A, U, V);
+        }
+    }
+};
+
+template <typename MatrixType>
+struct matrix_exp_computeUV<MatrixType, double> {
+    typedef typename NumTraits<typename traits<MatrixType>::Scalar>::Real RealScalar;
+    template <typename ArgType>
+    static void run(const ArgType& arg, MatrixType& U, MatrixType& V, int& squarings) {
+        using std::frexp;
+        using std::pow;
+        const RealScalar l1norm = arg.cwiseAbs().colwise().sum().maxCoeff();
+        squarings = 0;
+        if (l1norm < 1.495585217958292e-002) {
+            matrix_exp_pade3(arg, U, V);
+        } else if (l1norm < 2.539398330063230e-001) {
+            matrix_exp_pade5(arg, U, V);
+        } else if (l1norm < 9.504178996162932e-001) {
+            matrix_exp_pade7(arg, U, V);
+        } else if (l1norm < 2.097847961257068e+000) {
+            matrix_exp_pade9(arg, U, V);
+        } else {
+            const RealScalar maxnorm = 5.371920351148152;
+            frexp(l1norm / maxnorm, &squarings);
+            if (squarings < 0)
+                squarings = 0;
+            MatrixType A = arg.unaryExpr(MatrixExponentialScalingOp<RealScalar>(squarings));
+            matrix_exp_pade13(A, U, V);
+        }
+    }
+};
+
+template <typename MatrixType>
+struct matrix_exp_computeUV<MatrixType, long double> {
+    template <typename ArgType>
+    static void run(const ArgType& arg, MatrixType& U, MatrixType& V, int& squarings) {
+#if LDBL_MANT_DIG == 53  // double precision
+        matrix_exp_computeUV<MatrixType, double>::run(arg, U, V, squarings);
+
+#else
+
+        using std::frexp;
+        using std::pow;
+        const long double l1norm = arg.cwiseAbs().colwise().sum().maxCoeff();
+        squarings = 0;
+
+#if LDBL_MANT_DIG <= 64  // extended precision
+
+        if (l1norm < 4.1968497232266989671e-003L) {
+            matrix_exp_pade3(arg, U, V);
+        } else if (l1norm < 1.1848116734693823091e-001L) {
+            matrix_exp_pade5(arg, U, V);
+        } else if (l1norm < 5.5170388480686700274e-001L) {
+            matrix_exp_pade7(arg, U, V);
+        } else if (l1norm < 1.3759868875587845383e+000L) {
+            matrix_exp_pade9(arg, U, V);
+        } else {
+            const long double maxnorm = 4.0246098906697353063L;
+            frexp(l1norm / maxnorm, &squarings);
+            if (squarings < 0)
+                squarings = 0;
+            MatrixType A = arg.unaryExpr(MatrixExponentialScalingOp<long double>(squarings));
+            matrix_exp_pade13(A, U, V);
+        }
+
+#elif LDBL_MANT_DIG <= 106  // double-double
```

Comment:
I don't think existing nmodl supports this much precision; is this just for internal calculations? Or are you future-proofing for high precision?

Reply:
This file was copied from the eigen library so it has extra stuff which we won't need.

I copied this file from `nrn/external/eigen/unsupported/`. Then I modified this file, by adding the `#undef` statements and by inlining the function `stem_function_exp` which was previously located in a separate header.